packages/gatsby/src/internal-plugins/load-babel-config/gatsby-node.js

```diff
@@ -0,0 +1,194 @@
+  const babelrcPath = path.join(directory, `.babelrc.js`)
+  try {
+    babelrc = preferDefault(require(babelrcPath))
+  } catch (error) {
+    if (!testRequireError(babelrcPath, error)) {
+      throw error
+    }
+  }
+
+  // TODO support this
+  if (typeof babelrc === `function`) {
+    report.error(
+      `.babelrc.js files that export a function are not supported in Gatsby`
+    )
+    return null
+  }
+
+  return babelrc
+}
+
+/**
+ * Reads the user's package.json and returns the "babel" section. It will
+ * return undefined when the "babel" section does not exist.
+ */
+function findBabelPackage(directory) {
+  const packageJson = require(path.join(directory, `package.json`))
+  try {
+    // $FlowFixMe - https://github.com/facebook/flow/issues/1975
+    return packageJson.babel
+  } catch (error) {
+    if (testRequireError(packageJson, error)) {
+      return null
+    } else {
+      throw error
+    }
+  }
+}
+
+/**
+ * Convert a babelrc from a .babelrc file or package.json
+ * to actions to add it to the store.
+ */
+const actionifyBabelrc = (babelrc, actions) => {
+  const { presets, plugins, ...options } = babelrc
+  if (presets && _.isArray(presets)) {
+    presets.forEach(p => {
+      let name
+      let options = {}
+      if (_.isArray(p)) {
+        name = p[0]
+        options = p[1]
+      } else {
+        name = p
+      }
+      actions.setBabelPreset({
+        name,
+        options,
+      })
+    })
+  }
+  if (plugins && _.isArray(plugins)) {
+    plugins.forEach(p => {
+      let name
+      let options = {}
+      if (_.isArray(p)) {
+        name = p[0]
+        options = p[1]
+      } else {
+        name = p
+      }
+      actions.setBabelPlugin({
+        name,
+        options,
+      })
+    })
+  }
+
+  if (_.isObject(options) && !_.isEmpty(options)) {
+    actions.setBabelOptions(options)
+  }
+}
+
+exports.actionifyBabelrc = actionifyBabelrc
+
+/**
+ * Creates a normalized Babel config to use with babel-loader. Loads a local
+ * babelrc config if one exists or sets a backup default.
+ */
+exports.onCreateBabelConfig = ({ stage, store, actions }) => {
+  const program = store.getState().program
+  const { directory } = program
+
+  let babelrc =
+    findBabelrc(directory) ||
+    findBabelrcJs(directory) ||
+    findBabelPackage(directory)
+
+  // If user doesn't have a custom babelrc, add defaults.
+  if (babelrc) {
+    actionifyBabelrc(babelrc, actions)
+  } else {
+    addDefaultPluginsPresets(actions, {
+      stage,
+      browserslist: program.browserslist,
+    })
+  }
+}
+
+const addDefaultPluginsPresets = (
+  actions,
+  { stage = ``, browserslist = {} }
+) => {
+  // Presets
+  actions.setBabelPreset({
+    name: `@babel/preset-env`,
+    stage,
+    options: {
+      loose: true,
+      modules: false,
+      useBuiltIns: `usage`,
+      sourceType: `unambiguous`,
+      shippedProposals: true, // includes async/await and Object spread/rest
+      targets: { browsers: browserslist },
```

Comment:
I learned the hard way the other day that `preset-env` should always force include `object-rest-spread` (https://github.com/4Catalyzer/babel-preset-4catalyzer/blob/master/index.js#L41) webpack can't handle the syntax even if the browser can.

Reply:
Huh, that sounds really painful to debug.

`shippedProposals` does add this — perhaps we should validate user's babelrcs to make sure they've either set `shippedPropsals` or that they include it directly.